nixos: add sharedModules and extraSpecialArgs options

Pacman99 opened the pull request:
### Description

<!--

Please provide a brief description of your change.

-->

closes #1698

This allows users of the nixos and nix-darwin module to set shared modules for all users and extra specialArgs to be available to home-manager modules.

The latter is named extraSpecialArgs just like the argument to modules/default.nix. This could be confusing since the the two are independent in code, but they do mean the same thing so I think the name fits.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

thiagokokada reviewed (approved):
I was going to open this exactly PR, thank you.

LGTM.

thiagokokada commented:
BTW, maybe we could add something similar to nix-darwin so we can finally close issue #1698.

Pacman99 commented:
It looks like half of this PR is already done at #1757 so I can remove my extraModules option if that one gets merged.

Pacman99 commented:
I copied over the changes to the nix-darwin module too along with the nixosConfig specialArg.

thiagokokada reviewed (approved):
LGTM.

thiagokokada commented:
@rycee @berbiche Could we have this PR reviewed? It makes some use cases in HM with Flakes much easier.

I am using this PR in my[ backported version HM](https://github.com/thiagokokada/home-manager/commit/eda1b85be3ddbb553b08349d574a8982f788ece3) with great success (well, at least the `extraSpecialArgs` part, since I don't have usage for `extraModules` yet)..

berbiche commented on nix-darwin/default.nix:
```diff
@@ -9,7 +9,10 @@ let
-    specialArgs = { lib = extendedLib; };
+    specialArgs = { 
+      lib = extendedLib; 
+      nixosConfig = config;
```
```suggestion
      darwinConfig = config;
```

Feels more appropriate.

berbiche reviewed (changes requested):
Thanks for the contribution.
Few things to fix.
I'm not sure of the exact syntax of docbook used in `description`s.

Pacman99 commented:
Thanks for the review, fixed and rebased.

> I am using this PR in my backported version HM with great success (well, at least the extraSpecialArgs part, since I don't have usage for extraModules yet).
I'm glad it works! I think the extraSpecialArgs is more useful for me too. I find the extra modules useful to prevent repeating code for multiple users. And I want it for nixflk, so we can automatically wire private modules.

berbiche commented on nix-darwin/default.nix:
```diff
@@ -55,6 +58,24 @@ in
+      extraModules = mkOption {
+        type = with types; listOf (oneOf [ attrs (functionTo attrs) path ]);
+        default = [ ];
+        example = "[ { home.packages = [ nixpkgs-fmt ]; } ]";
```
```suggestion
        example = literalExample "[ { home.packages = [ nixpkgs-fmt ]; } ]";
```

berbiche commented on nix-darwin/default.nix:
```diff
@@ -55,6 +58,24 @@ in
+      extraSpecialArgs = mkOption {
+        type = types.attrs;
+        default = { };
+        example = "{ modulesPath = ../modules; }";
```
```suggestion
        example = literalExample "{ modulesPath = ../modules; }";
```

berbiche commented on nixos/default.nix:
```diff
@@ -65,6 +65,24 @@ in {
+      extraSpecialArgs = mkOption {
+        type = types.attrs;
+        default = { };
+        example = "{ modulesPath = ../modules; }";
```
```suggestion
        example = literalExample "{ modulesPath = ../modules; }";
```

berbiche commented on nixos/default.nix:
```diff
@@ -65,6 +65,24 @@ in {
+      extraModules = mkOption {
+        type = with types; listOf (oneOf [ attrs (functionTo attrs) path ]);
+        default = [ ];
+        example = "[ { home.packages = [ nixpkgs-fmt ]; } ]";
```
```suggestion
        example = literalExample "[ { home.packages = [ nixpkgs-fmt ]; } ]";
```

berbiche reviewed (changes requested):
Sorry, I missed a few thing earlier.

Pacman99 commented:
Cool fixed that. Also I squashed the commits for each file, so I added both options in one commit. Since their pretty related, and its easier to change the options. But I can split it again if thats preferred.

berbiche commented on nix-darwin/default.nix:
```diff
@@ -55,6 +58,24 @@ in
+        '';
+      };
+
+      extraModules = mkOption {
+        type = with types; listOf (oneOf [ attrs (functionTo attrs) path ]);
+        default = [ ];
+        example = literalExample "[ { home.packages = [ nixpkgs-fmt ]; } ]";
+        description = ''
+          Extra modules added to all users.
+        '';
+      };
```
```suggestion
      extraDefaultModules = mkOption {
        type = with types; listOf (oneOf [ attrs (functionTo attrs) path ]);
        default = [ ];
        example = literalExample "[ { home.packages = [ nixpkgs-fmt ]; } ]";
        description = ''
          Extra modules added to all users.
        '';
      };
```

What do you think of this name?

Same for the one used in `nixos/default.nix`.

Pacman99 replied:
I don't mind it, but its a little confusing to me. Default usually means this option is the default *to* something. But these are the only modules set and all of them get added. I'm not entirely against it, but it feels a bit weird. 

I do like the name the other PR used, sharedModules.

berbiche replied:
Let's go for `sharedModules` then

Pacman99 replied:
sounds good, changed it

rycee commented:
The `nix-darwin: add darwinConfig specialArg` doesn't look correct. Maybe some squashing mistake?

rycee commented on nix-darwin/default.nix:
```diff
@@ -9,7 +9,11 @@ let
+      lib = extendedLib; 
+      darwinConfig = config;
+    };
+    } // cfg.extraSpecialArgs;
```
Maybe this line is intended for the next commit?

Pacman99 replied:
Ohh right it was a hiccup while trying to split up the changes with `git add --patch`. Just fixed, thanks for the catch!
